modules/ldapbackend/dnssec.cc

```diff
@@ -0,0 +1,1386 @@
+  {
+    L << Logger::Error << d_myname << " Caught STL exception setting metadata for domain " << name << ": " << e.what() << endl;
+    throw DBException( "STL exception" );
+  }
+}
+
+
+bool LdapBackend::getDomainKeys( const DNSName& name, std::vector<KeyData>& keys )
+{
+  if ( !d_dnssec )
+    return false;
+
+  std::string basedn = this->getDomainMetadataDN( name );
+  if ( basedn.empty() )
+    return false;
+
+  L<<Logger::Debug<< d_myname << " Retrieving zone keys for " << name << std::endl;
+
+  try {
+    PowerLDAP::sentry_t result;
+    std::string filter = "objectClass=PdnsDomainKey";
+    const char* attributes[] = {
+      "PdnsKeyId",
+      "PdnsKeyFlags",
+      "PdnsKeyContent",
+      "PdnsKeyActive",
+      NULL
+    };
+
+    PowerLDAP::SearchResult::Ptr search = d_pldap->search( basedn, LDAP_SCOPE_SUBTREE, filter, attributes );
+
+    while ( search->getNext( result, false ) ) {
+      KeyData kd;
+      kd.id = pdns_stou( result["PdnsKeyId"][0] );
+      kd.flags = pdns_stou( result["PdnsKeyFlags"][0] );
+      kd.content = result["PdnsKeyContent"][0];
+      if ( result.count( "PdnsKeyActive" ) )
+        kd.active = true;
+      else
+        kd.active = false;
+
+      L<<Logger::Debug<< d_myname << " Found key with ID " << kd.id << std::endl;
+      keys.push_back( kd );
+    }
+
+    return true;
+  }
+  catch( LDAPTimeout &lt )
+  {
+    L << Logger::Warning << d_myname << " Unable to search LDAP directory: " << lt.what() << endl;
+    throw DBException( "LDAP server timeout" );
+  }
+  catch( LDAPNoConnection &lnc )
+  {
+    L << Logger::Warning << d_myname << " Connection to LDAP lost, trying to reconnect" << endl;
+    if ( reconnect() )
+      return this->getDomainKeys( name, keys );
+    else
+      throw PDNSException( "Failed to reconnect to LDAP server" );
+  }
+  catch( LDAPException &le )
+  {
+    L << Logger::Error << d_myname << " Unable to search LDAP directory: " << le.what() << endl;
+    throw PDNSException( "LDAP server unreachable" );   // try to reconnect to another server
+  }
+  catch( std::exception &e )
+  {
+    L << Logger::Error << d_myname << " Caught STL exception retrieving key for domain " << name << ": " << e.what() << endl;
+    throw DBException( "STL exception" );
+  }
+}
+
+
+bool LdapBackend::addDomainKey( const DNSName& name, const KeyData& key, int64_t& id )
+{
+  if ( !d_dnssec )
+    return false;
+
+  L<<Logger::Debug<< d_myname << " Adding domain key for " << name << std::endl;
+
+  // Same as in setDomainMetadata, we don't create the required entries
+  std::string basedn = this->getDomainMetadataDN( name );
+  if ( basedn.empty() )
+    return false;
+
+  try {
+    // Prepare all the elements first. As LDAP doesn't have the concept of transaction
+    // we risk ending up with two keys with the same ID if there are insertions between
+    // the moment we get the list of keys and the moment we insert. So better not dawdle
+    // between those.
```

Comment:
This seems hacky, just shortening the race window. Is there no way we can do better than this?